File: BlitzBi/Module/BlitzBiEventRepository.m
//
//  BlitzBiEventRepository.m
//  BlitzBi
//
//  Created by Admin on 12/10/20.
//

#import <BlitzBiEventRepository.h>

@implementation BlitzBiEventRepository
- (instancetype)init:(NSNumber*)appId
                    :(NSString*)appToken
                    :(id <PBlitzDataTransferService>)networkService {
    if (self = [super init]) {
        self->appId = appId;
        self->appToken = appToken;
        self->networkService = networkService;
    }
    return self;
}
- (void)processJsonRequest:(NSString*)url
                          :(NSData*)data
                          :(void(^)(NSObject *, NSError *))completion{
    BlitzRequestBuilder *requestBuilder = [[BlitzRequestBuilder alloc] init];
    [requestBuilder setMethod:POST_METHOD];
    [requestBuilder setBaseUrl:url];
    [requestBuilder setPath:@"events"];
    [requestBuilder setHeaders:[self getRequestHeaders]];
    [requestBuilder setHttpBody:data];
    [requestBuilder setReqType:BI_REQUEST];
    [requestBuilder setContentType:@"application/json"];
    
    [networkService executeServerCall:requestBuilder withCompletion: completion];
}

- (void)processJsonRequestWithoutResponse:(NSString*)url
                                         :(NSData*)data
                                         :(BOOL)isEmergency {
    BlitzRequestBuilder *requestBuilder = [[BlitzRequestBuilder alloc] init];
    [requestBuilder setMethod:POST_METHOD];
    [requestBuilder setBaseUrl:url];
    [requestBuilder setPath:@"events"];
    [requestBuilder setHeaders:[self getRequestHeaders]];
    [requestBuilder setHttpBody:data];
    [requestBuilder setReqType:BI_REQUEST];
    [requestBuilder setContentType:@"application/json"];
    
    if (isEmergency && ![NSThread isMainThread]) {
        [[[NSURLSession sharedSession] dataTaskWithRequest:[requestBuilder generateRequest] completionHandler:^(NSData *data, NSURLResponse *response, NSError *error){
            
        }] resume];
    } else {
        [networkService executeServerCallWithNoCallBack:requestBuilder];
    }
}

- (NSMutableDictionary*) getRequestHeaders {
    NSMutableDictionary *headers = [[NSMutableDictionary alloc] init];
    [headers setValue:[appId stringValue] forKey:@"blitzAppId"];
    [headers setValue:appToken forKey:@"blitzAppToken"];
    return headers;
}
@end

File: BlitzBi/Module/BlitzBiEventRepository.swift
//
//  BIEventRepository.swift
//  Bi
//
//  Created by Saurabh on 16/09/20.
//

import Foundation

protocol IBlitzBIEventRepository {
    func processJsonRequestWithoutResponse(url:String, body:Data?, isEmergency:Bool)
    func processJsonRequest(url:String, data:Data?, completionHandler:@escaping(BiResponse?,Error?) -> Void)
}

class BlitzBIEventRepository:IBlitzBIEventRepository {
    private var networkService:PBlitzDataTransferService
    private var appId: Int
    private var appToken: String
    
    init(networkService: PBlitzDataTransferService, appId: Int, appToken: String) {
        self.networkService = networkService
        self.appId = appId
        self.appToken = appToken
    }
    
    func processJsonRequest(url:String, data:Data?, completionHandler:@escaping(BiResponse?,Error?) -> Void) {
        let requestBuilder = BlitzRequestBuilder()
        requestBuilder.method = "POST"
        requestBuilder.baseUrl = url
        requestBuilder.path = "events"
        requestBuilder.contentType = "application/json"
        requestBuilder.httpBody = data
        requestBuilder.reqType = BI_REQUEST
        requestBuilder.headers = getRequestHeaders()

        networkService.executeServerCall(requestBuilder: requestBuilder, type: BiResponse.self, completionBlock: {result in
          switch(result) {
          case .success(let result):
              completionHandler(result,nil)
          case .failure(let error):
              completionHandler(nil,error)
          }
        })
      }
    
    func getRequestHeaders() -> NSMutableDictionary {
        let headers: NSMutableDictionary = [
            "blitzAppId" : String(appId),
            "blitzAppToken" : appToken
        ]
        return headers
    }
    
    func processJsonRequestWithoutResponse(url:String, body:Data?, isEmergency:Bool) {
        let requestBuilder = BlitzRequestBuilder()
        requestBuilder.method = "POST"
        requestBuilder.baseUrl = url
        requestBuilder.path = "events"
        requestBuilder.contentType = "application/json"
        requestBuilder.httpBody = body
        requestBuilder.reqType = BI_REQUEST
        requestBuilder.headers = getRequestHeaders()
        requestBuilder.responseListener = nil
        if (isEmergency  && !Thread.isMainThread) {
          do {
              try NSURLConnection.sendSynchronousRequest(requestBuilder.generateRequest() as URLRequest, returning: nil)
          } catch {
          }
        } else {
          networkService.executeServerCallWithNoCallBack(requestBuilder:requestBuilder)
        }
    }
}
